fs: Port of NFFS (Newtron Flash File System)

andrzej-kaczmarek opened the pull request:
This is port of NFFS (https://mynewt.apache.org/os/modules/fs/nffs/nffs/) to Zephyr. NFFS code is pulled in directly from https://github.com/apache/mynewt-nffs repository.

As opposed to Mynewt, in Zehypr it is configured not to use heap memory so everything is statically allocated here.

fs_truncate() call returns -ENOTSUP at the moment since there is no proper way to do this using current NFFS implementation.

Since NFFS is a *flash* filesystem, it does access flash directly instead of intermediate layer like disk subsystem (as FAT does). Flash area available for NFFS is defined in DTS and this is updated for nRF5x DKs from Nordic.

Test cases are also ported from Mynewt - the scenarios are almost the same (minor changes to cover differences between Mynewt and Zephyr FS API) but they run in Zephyr test framework.

andrzej-kaczmarek committed:
fs: nffs: Add NFFS filesystem support

This patch adds filesystem interface implementation for NFFS.
Default configuration for mem slabs sizes are the same as in Mynewt.

Origin: Original
Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
ext: nffs: Pull in Apache Mynewt NFFS implementation

Origin: Apache Mynewt NFFS
URL: https://github.com/apache/mynewt-nffs/tree/master
Commit: 6639f7a432e70db00ef25d5e558aedbe503a3c9a
Purpose: Introduction of NFFS (Newtron Flash File System)
Maintained-by: External

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

kolodgrz committed:
tests: nffs_fs_api: Add ported NFFS test suite

This patch adds NFFS test suite ported from Mynewt. The test cases are
pretty much the same as in original implementation, however some of
them had to be adjusted to cover differences in Mynewt and Zephyr
filesystem APIs. Also the code was reformatted in some places to make
checkpatch happy.

All test cases from "basic" group can be also run on nRF52840 (nRF52 and
nRF51 do not have enough RAM to run all tests as-is).

Origin: Apache Mynewt NFFS
URL: https://github.com/apache/mynewt-nffs/tree/master
Commit: 6639f7a432e70db00ef25d5e558aedbe503a3c9a
Purpose: NFFS test suite
Maintained-by: Zephyr

Signed-off-by: Grzegorz Kolodziejczyk <grzegorz.kolodziejczyk@codecoup.pl>
Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
dts: nrf5x: Setup partitions for NFFS

This patch adds partition for NFFS for nRF51, nRF52 and nRF52840.
The partition is placed at the end of flash. This is only added if NFFS
is enabled (since it's required) - in other case free space can be used
for other purposes.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
fs: shell: Add more shell functions

Add following commands to fs shell:
- rf <filename>
- read <filename> [<offset> [<length>]]
- write <filename> [-o <offset>] <byte1> [<byte2> ... [<byteN>]]
- trunc <filename> [<length>]

For writes without offset specified, new data is appended to file.
For truncaces without length specified, 0 is default value.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

carlescufi reviewed (changes requested):
I think the changes made to NFFS itself should be included in a future NFFS repo. Will follow-up with Runtime to create the repo

andrzej-kaczmarek commented:
@carlescufi yes, they will be part of NFFS repo and then complete code will be copied in single commit - for now I'll just keep this as separate commit here so changes are clearly visible and I can then easily move them to NFFS repo once it's created

kolodgrz committed:
tests: nffs_fs_api: Add performance test for nffs

This adds performance test for nffs with open, close, read, write and
unlink sub test cases.

Origin: Original

Signed-off-by: Grzegorz Kolodziejczyk <grzegorz.kolodziejczyk@codecoup.pl>
Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

nashif commented on subsys/fs/shell.c:
```diff
@@ -46,6 +54,204 @@ static int cmd_mkdir(int argc, char *argv[])
 	return 0;
 }
 
+static int cmd_rf(int argc, char *argv[])
```
why not rm? this is more consistent with the other command you are using.

andrzej-kaczmarek replied:
right, I'll change this

nashif commented on tests/subsys/fs/nffs_fs_api/Makefile:
```diff
@@ -0,0 +1,25 @@
+	CFLAGS += -DTEST_FLASH_OFFSET=0x20000
+endif
+
+BOARD ?= qemu_x86
```
move this to the top, otherwise the conditional above wont work

nashif commented on tests/subsys/fs/nffs_fs_api/src/nffs_test_utils.h:
```diff
@@ -0,0 +1,103 @@
+/*
```
any reason why you are including the full license text?

andrzej-kaczmarek replied:
most files in tests/ are files from NFFS repository adjusted to use with Zephyr so I left licenses intact (like for a NFFS code pull-in) - just added spdx

nashif reviewed (changes requested):
if you do not want the commit marked with [donotmerge] to be merged, then please remove it from the PR

andrzej-kaczmarek commented:
Extra commits for testing are now on nffs-dev-testing branch

nvlsianpu commented on ext/fs/nffs/include/nffs/nffs.h:
```diff
@@ -0,0 +1,544 @@
+                                                supersedes lesser. */
+    struct nffs_inode_entry *nb_inode_entry; /* Owning inode. */
+    struct nffs_hash_entry *nb_prev;         /* Previous block in file. */
+    uint16_t nb_data_len;                    /* # of data bytes in block. */
```
inconsistent slang is used (# vs # off line195). Personally I prefer to avoid slang in comment.

andrzej-kaczmarek replied:
the whole ext/fs/nffs/ code is pulled-in for external repository and I'd prefer to avoid any extra changes here since this code is meant to be maintained outside Zephyr - otherwise there is plenty of other stuff to adjust to Zephyr standards here and we'll quickly diverge from the original code.

carlescufi commented on dts/arm/nrf52_pca10040.dts:
```diff
@@ -24,3 +24,49 @@
 	current-speed = <115200>;
 };
+
+&flash0 {
```
Have you rebased against the latest master? There is already a flash entry in this file in master

carlescufi replied:
my mistake

carlescufi commented:
@andrzej-kaczmarek I think you need to rebase your patches against the latest master

nvlsianpu commented on ext/fs/nffs/include/nffs/os.h:
```diff
@@ -0,0 +1,70 @@
+ * under the License.
+ */
+
+#ifndef H_GLUE_
```
Include fuse and file name should stick somehow. I suggest to change the file name to something more descriptive ( `os_glue.h', nffs_os_glue.h' ...) and to reflect this here.

andrzej-kaczmarek replied:
I'll change this #ifdef in nffs reposity since it's an obvious mistake there

carlescufi commented on subsys/fs/nffs_fs.c:
```diff
@@ -0,0 +1,534 @@
+#include <nffs/nffs.h>
+#include <nffs/os.h>
+
+static struct k_mutex nffs_lock;
```
Can you add a comment here elaborating on the need for this mutex? In particular how the NFFS code is not thread safe and requires locking outside of its calls

carlescufi replied:
thanks!

nvlsianpu commented on ext/fs/nffs/include/nffs/nffs.h:
```diff
@@ -0,0 +1,544 @@
+#define    NFFS_INODE_FLAG_INHASH      0x40    /* in hash table */
+#define    NFFS_INODE_FLAG_DELETED     0x80    /* inode deleted */
+
+#define nie_id            nie_hash_entry.nhe_id
```
`nie_id` is unused

carlescufi commented:
@andrzej-kaczmarek CI check are failing because of checkpatch, you can find more details here: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/4321/3/tests

nvlsianpu commented on ext/fs/nffs/include/nffs/queue.h:
```diff
@@ -0,0 +1,522 @@
+ * A circle queue may be traversed in either direction, but has a more
+ * complex end of list detection.
+ *
+ * For details on the use of these macros, see the queue(3) manual page.
```
queue(3) is probably well known for someone who is familiar with linux (so not for me), There are a few manuals so it is worth to mention a one? Not necessary by the weebpage, but rather "for instance free bsd lib. manual"

nvlsianpu replied:
In case you want keep this file as original - its still ok for me.

nvlsianpu commented on ext/fs/nffs/src/nffs_area.c:
```diff
@@ -0,0 +1,117 @@
+static void
+nffs_area_set_magic(struct nffs_disk_area *disk_area)
+{
+    disk_area->nda_magic[0] = NFFS_AREA_MAGIC0;
```
Better is to declare a constant array of `the magic value` and use it here with memcpy.

nvlsianpu commented on ext/fs/nffs/src/nffs_area.c:
```diff
@@ -0,0 +1,117 @@
+int
+nffs_area_magic_is_set(const struct nffs_disk_area *disk_area)
+{
+    return disk_area->nda_magic[0] == NFFS_AREA_MAGIC0 &&
```
Replace this using memcmp.

nvlsianpu commented on ext/fs/nffs/src/nffs_area.c:
```diff
@@ -0,0 +1,117 @@
+nffs_area_to_disk(const struct nffs_area *area,
+                  struct nffs_disk_area *out_disk_area)
+{
+    memset(out_disk_area, 0, sizeof *out_disk_area);
```
(performance) Is it worth to clear whole structure and then fill it again? I suggest just to clear the reserved field.

andrzej-kaczmarek commented:
* updated NFFS code to match latest commit in mynewt-nffs repository
* numbers for dts partitions are updated to match https://github.com/zephyrproject-rtos/zephyr/pull/1307
* tests/ code was reformatted a bit to make checkpatch happy

carlescufi commented on dts/arm/nrf51_pca10028.dts:
```diff
@@ -23,3 +23,49 @@
+		};
+#else
+		/*
+		 * The flash starting at 0x0003e000 and ending at
```
This comment should be moved up and reflect the fact that NFFS occupies the flash area described in the comment. Same for the other files.

nvlsianpu commented on include/fs/fs_interface.h:
```diff
@@ -43,6 +43,8 @@ extern "C" {
 
 #ifdef CONFIG_FILE_SYSTEM_FAT
```
Are there any blocker for support FAT and NFFS simultaneously?

andrzej-kaczmarek replied:
both fat_fs.c and nffs_fs.c will export the same symbols for fs api so it won't link

nvlsianpu replied:
Are there any other obstacles? 

I think about cases like - NFFS for internal storage (e.g. BLE bond data) and FAT for external SD card. For sure  I understand that this is not in scope of this PR intention.

nashif replied:
we do need to address this, there is a JIRA about this

carlescufi commented on subsys/fs/Kconfig:
```diff
@@ -22,12 +22,61 @@ config FILE_SYSTEM_SHELL
+menu "NFFS Settings"
+	visible if FILE_SYSTEM_NFFS
+
+config FILE_SYSTEM_NFFS_FLASH_DEV_NAME
```
I suggest we rename FILE_SYSTEM_NFFS_* to FS_NFFS_*. We made this mistake in Bluetooth and it quickly becomes too long.

nvlsianpu commented on ext/fs/nffs/Kconfig:
```diff
@@ -0,0 +1,17 @@
+	help
+	Use the NFFS File system implementation.
+
+config NFFS_FILESYSTEM_MAX_AREAS
```
can we have a help here and below. What is it? How these values impose on the file system(performance, corner cases etc.).

nvlsianpu commented on subsys/fs/nffs_fs.c:
```diff
@@ -0,0 +1,540 @@
+		return rc;
+	}
+
+	rc = flash_write(flash_dev, address, src, num_bytes);
```
After the write flash may be leaved unprotected for some device (for instance all nrf5x SoC). Can you add the lock.

nvlsianpu commented on subsys/fs/nffs_fs.c:
```diff
@@ -0,0 +1,540 @@
+	}
+
+	rc = flash_erase(flash_dev, address, num_bytes);
+
```
After the write flash may be leaved unprotected for some devic... see coment above.

nvlsianpu commented on subsys/fs/nffs_fs.c:
```diff
@@ -0,0 +1,540 @@
+	struct flash_pages_info pi;
+	int rc;
+
+	rc = flash_get_page_info_by_idx(flash_dev, sector, &pi);
```
FLASH_PAGE_LAYOUT flash driver option is used. Add force selection of FLASH_PAGE_LAYOUT in case NFFS is selected.

carlescufi replied:
Please don't use "depends" and instead use "select"

nvlsianpu commented on subsys/fs/nffs_fs.c:
```diff
@@ -0,0 +1,540 @@
+
+	rc = nffs_path_find(&parser, &inode, &parent);
+	if (rc) {
+		goto done;
```
`goto` dose not fit to any style in this file and it is too trivial case for use it, rearange thi to if (!rc) ...

nvlsianpu commented on subsys/fs/shell.c:
```diff
@@ -46,6 +54,205 @@ static int cmd_mkdir(int argc, char *argv[])
+		return 0;
+	}
+
+	if (argv[1][0] == '/') {
```
Please extract this code block as it repeats a few times in this file.

nvlsianpu commented on subsys/fs/shell.c:
```diff
@@ -46,6 +54,205 @@ static int cmd_mkdir(int argc, char *argv[])
+		buf[buf_len++] = strtol(argv[arg_offset++], NULL, 16);
+
+		if ((buf_len == BUF_CNT) || (arg_offset == argc)) {
+			fs_write(&file, buf, buf_len);
```
fs_write might returns failure - please add appropriate error handling.

andrzej-kaczmarek committed:
fs: shell: Extract common path handling code to helper

This patch moves code to create absolute path of file/directory to
separate helper since the same code is used in few places.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented:
Updated with requested fixes.

andrzej-kaczmarek commented:
Added missing signed-off-by

nvlsianpu commented:
Thanks for changes! Now it looks good.
